docs/source/tutorial_optimization.rst

```diff
@@ -0,0 +1,900 @@
+        data['evaluations'].append(QoI[0])
+        opt_dict['opt_rom'] = opt_rom 
+        return QoI
+    
+    def compute_gradient_with_opt_rom(opt_dict, mu):
+        opt_rom = opt_dict['opt_rom']
+        return opt_rom.output_d_mu(opt_rom.parameters.parse(mu), return_array=True, use_adjoint=True)
+
+.. jupyter-execute::
+
+    opt_along_path_adaptively_minimization_data = {'num_evals': 0,
+                                           'evaluations' : [],
+                                           'evaluation_points': [],
+                                           'time': np.inf,
+                                           'enrichments': 0}
+    opt_dict = {'opt_rom': opt_rom}
+    tic = perf_counter()
+    opt_along_path_adaptively_result = minimize(partial(record_results_and_enrich_adaptively, rom_objective_functional, 
+                                                        opt_along_path_adaptively_minimization_data, opt_dict),
+                                                initial_guess.to_numpy(),
+                                                method='L-BFGS-B', 
+                                                jac=partial(compute_gradient_with_opt_rom, opt_dict),
+                                                bounds=(ranges, ranges),
+                                                options={'ftol': 1e-15})
+    opt_along_path_adaptively_minimization_data['time'] = perf_counter()-tic
+
+
+
+.. jupyter-execute::
+
+    report(opt_along_path_adaptively_result, opt_along_path_adaptively_minimization_data, reference_mu)
+
+    
+
+
+Now, we actually only needed :math:`4` enrichments and ended up with an
+approximation error of about ``1e-07`` while getting the highest speed up
+amongst all methods that we have seen above. To conclude, we once again
+compare all methods that we have discussed in this notebook.
+
+
+.. jupyter-execute::
+
+    print('FOM with finite differences')
+    report(fom_result, reference_minimization_data, reference_mu)
+    
+    print('\nROM with finite differences')
+    report(rom_result, RB_minimization_data, reference_mu)
+    
+    print('\nFOM with gradient')
+    report(opt_fom_result, opt_fom_minimization_data, reference_mu)
+    
+    print('\nROM with gradient')
+    report(opt_rom_result, opt_rom_minimization_data, reference_mu)
+    
+    print('\nAlways enrich along the path')
+    report(opt_along_path_result, opt_along_path_minimization_data, reference_mu)
+    
+    print('\nAdaptively enrich along the path')
+    report(opt_along_path_adaptively_result, opt_along_path_adaptively_minimization_data, reference_mu)
+
+.. jupyter-execute::
+    :hide-code:
+    :hide-output:
+    
+    assert fom_result.nit == 10
+    assert opt_along_path_result.nit == 8
+    assert opt_along_path_minimization_data['num_evals'] == 10
+    assert opt_along_path_minimization_data['enrichments'] == 9
+    assert opt_along_path_adaptively_minimization_data['enrichments'] == 4
+
+
```

Comment:
This still depends on the magic tolerance for enriching the model. Probably an easy way around this is to do one optimization step for the FOM after converging. If this changes anything, the ROM tolerance was too large. Probably too much work for the tutorial, but one might at least discuss the possibility.